### PR TITLE
feat(xrd): Add support for version deprecation

### LINKED ISSUE #2945: XRD versions do not support deprecation
### What happened?

XRDs do not support the `deprecated` and `deprecationWarning` fields that [are available for CRDs](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/).

### How can we reproduce it?

1. Define a XRD with a version. Set `spec.versions.deprecated` to true.
2. Run `kubectl apply`

### What environment did it happen in?

Crossplane version: 1.6.3

### PR DESCRIPTION
**Pull request by MisterMX**
### Description of your changes

This adds two new fields `deprecated` and `deprecationWarning` to XRDs `spec.versions[]` that match their CRD counterparts.

Fixes #2945 

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### How has this code been tested

Manually

### CODE REVIEW

File: apis/apiextensions/v1/xrd_types.go [hunk 1]
```diff
@@ -98,6 +98,14 @@ type CompositeResourceDefinitionVersion struct {
 	// Served specifies that this version should be served via REST APIs.
 	Served bool `json:"served"`
 
+	// The deprecated field specifies that this version is deprecated and should not be used.
+	// +optional
+	Deprecated *bool `json:"deprecated,omitempty"`
+
+	// DeprecationWarning specifies the message that should be shown to the user when using this version.
```
Comment (hasheddan) [inline apis/apiextensions/v1/xrd_types.go:105]: Are these comments wrapped at 80 characters? If not could you make sure to wrap them! 🙂

**Review by hasheddan — APPROVED** [review]
LGTM -- I'm a fan of this change even though in practice I believe we can still only offer one XRD version at a time. Would like to get @negz thoughts before merge.

File: apis/apiextensions/v1/xrd_types.go (hunk 1 above)
Comment (MisterMX) [reply apis/apiextensions/v1/xrd_types.go]: Fixed

**Comment by MisterMX** [comment]
@hasheddan thanks for the review! I fixed the line lengths. Let me know if something is missing.

Regarding serving multiple XRD versions: AFAIK Crossplane _does_ support multiple versions at the same time.

However, our main use case is to let users know when they use a deprecated version. In some cases we do deprecate complete XRDs so we might need this for every version.

**Comment by negz** [comment]
> AFAIK Crossplane does support multiple versions at the same time.

Kind of. It supports it as long as all the versions are identical. :) So, that said, this feature makes sense. e.g.:

1. You start with a v1alpha1 API.
2. You 'promote' the API by introducing an identical v1beta1 variant.
3. You mark the v1alpha1 API deprecated.

Eventually we'll support introducing new API versions with different schemas per https://github.com/crossplane/crossplane/issues/2608.

**Review by negz — APPROVED** [review]

**Pull request merged**